### PR TITLE
Add UT for ReconcileVPC and its subfunctions

### LINKED ISSUE #1818: EPIC: Add UT accross various files in repository
/kind feature
/area provider/ibmcloud

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Recently we added lots of functionality but haven't added any corresponding UT's to it. Its time to add unit test cases and overall goal is to reach 80% of test coverage.

As a non-goal during the time of adding UT if we find any redundant or unwanted or code duplication, a separate PR needs to be submitted to optimize the code.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

- [ ] cloud/scope/powervs_cluster.go 
- [ ] cloud/scope/powervs_machine.go 

Instructions for running individual tests on VS Code

1. Get kubebuilder asset path
```
$ cd cluster-api-provider-ibmcloud
$ make setup-envtest
kube-builder assets set using darwin OS at location /Users/karthikkn/Library/Application Support/io.kubebuilder.envtest/k8s/1.29.3-darwin-amd64
```

2. Create settings.json file with kubebuilder asset path
```
$ cd cluster-api-provider-ibmcloud/.vscode
$ cat settings.json
cat settings.json
{
    "go.testEnvVars": {
        "KUBEBUILDER_ASSETS": "/Users/karthikkn/Library/Application Support/io.kubebuilder.envtest/k8s/1.29.3-darwin-amd64",
    },
}

```
3. Run individual test by clicking on run test above test function name in VS Code editor

Instructions for checking code coverage of file 

1. Generate the test coverage html page
```
$ cd cluster-api-provider-ibmcloud/.vscode
$ make test-cover
```
A file named cover.html will be created under cluster-api-provider-ibmcloud directory

example path: /Users/karthikkn/k8-workspace/cluster-api-provider-ibmcloud/cover.html

2. Copy the path of the file and paste in browser to view the html page and select the file to view the lines covered in UT.

image

```[tasklist]
### Helper PRs
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1913
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1899
```

```[tasklist]
### UT PRs
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1912
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1927
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1928
```

### PR DESCRIPTION
**Pull request by Shilpa-Gokul**
Add UT for ReconcileVPC, createVPC, checkVPC, getVPCByName and GetServiceName functions
Partially fixes #1818

### CODE REVIEW

File: cloud/scope/powervs_cluster_test.go [hunk 1]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+	}
+	t.Run("When VPC already exists in cloud, ", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{
+				VPC:           &infrav1beta2.VPCResourceReference{},
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1730]: Is this required to define here?

File: cloud/scope/powervs_cluster_test.go [hunk 2]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When VPC already exists in cloud, ", func(t *testing.T) {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1722]:
```suggestion
	t.Run("When VPC already exists in cloud", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 3]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{
+				VPC:           &infrav1beta2.VPCResourceReference{},
+				ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{ID: ptr.To("resourceGroupID")}}},
+		}
+		vpcOutput := &vpcv1.VPC{Name: ptr.To("VPCName"), ID: ptr.To("VPCID")}
+		mockVPC.EXPECT().GetVPCByName(gomock.Any()).Return(vpcOutput, nil)
+
+		output, err := clusterScope.ReconcileVPC()
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1736]:
```suggestion
		requeue, err := clusterScope.ReconcileVPC()
```
(lets replace output with requeue wherever applicable)

File: cloud/scope/powervs_cluster_test.go [hunk 4]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		g.Expect(err).To(BeNil())
+		g.Expect(output).To(BeFalse())
+		g.Expect(clusterScope.IBMPowerVSCluster.Status.VPC.ID).To(Equal(vpcOutput.ID))
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1739]: lets check for controllercreated as well.

File: cloud/scope/powervs_cluster_test.go [hunk 5]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{
+				//VPC:           &infrav1beta2.VPCResourceReference{ID: ptr.To("VPCID")},
+				ResourceGroup: &infrav1beta2.IBMPowerVSResourceReference{ID: ptr.To("resourceGroupID")}},
+				Status: infrav1beta2.IBMPowerVSClusterStatus{VPC: &infrav1beta2.ResourceReference{ID: ptr.To("VPCID")}},
+			},
+		}
+		mockVPC.EXPECT().GetVPC(gomock.Any()).Return(vpcOutput, nil, nil)
+		output, err := clusterScope.ReconcileVPC()
+		g.Expect(err).To(BeNil())
+		g.Expect(output).To(BeTrue())
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1869]: Is validating vpc id set in status is also checked?

File: cloud/scope/powervs_cluster_test.go [hunk 6]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When resourceGroupID is nil ", func(t *testing.T) {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1887]:
```suggestion
	t.Run("When resourceGroupID is nil", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 7]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient: mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{
+				VPC: &infrav1beta2.VPCResourceReference{}}},
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1895]: is this required?

File: cloud/scope/powervs_cluster_test.go [hunk 8]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient:      mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{ObjectMeta: metav1.ObjectMeta{Name: "ClusterName"}},
+		}
+		mockVPC.EXPECT().GetVPCByName(gomock.Any()).Return(nil, fmt.Errorf("GetVPCByName returns error"))
+		output, err := clusterScope.getVPCByName()
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(output).To(BeNil())
+	})
+	t.Run("When GetVPCByName returns proper vpc details", func(t *testing.T) {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:2124]: may be instead of proper can we use "valid"

File: cloud/scope/powervs_cluster_test.go [hunk 9]
```diff
@@ -1704,3 +1704,515 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient:      mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{ObjectMeta: metav1.ObjectMeta{Name: "ClusterName"}},
+		}
+		mockVPC.EXPECT().GetVPCByName(gomock.Any()).Return(nil, fmt.Errorf("GetVPCByName returns error"))
+		output, err := clusterScope.getVPCByName()
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:2120]: lets avoid using generic term output wherever applicable and use specific term related return values like vpcResponse ..

File: cloud/scope/powervs_cluster_test.go (hunk 1 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:1730]: Its not required, removed all such references

File: cloud/scope/powervs_cluster_test.go (hunk 3 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:1736]: Sure, have changed it

File: cloud/scope/powervs_cluster_test.go (hunk 4 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:1739]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 5 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:1869]: There is no status set [here](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/cloud/scope/powervs_cluster.go#L1004), should i check for nil instead?

File: cloud/scope/powervs_cluster_test.go (hunk 8 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:2124]: Done

File: cloud/scope/powervs_cluster_test.go (hunk 9 above)
Comment (Shilpa-Gokul) [reply cloud/scope/powervs_cluster_test.go:2120]: Sure, renamed it to proper terms everywhere

File: cloud/scope/powervs_cluster_test.go (hunk 5 above)
Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go]: oh ok, not sure why we are not setting status, In other resources we are setting status right, when spe.id is set?

File: cloud/scope/powervs_cluster_test.go [hunk 10]
```diff
@@ -1704,3 +1704,521 @@ func TestCreateServiceInstance(t *testing.T) {
...
+				Status: infrav1beta2.IBMPowerVSClusterStatus{VPC: &infrav1beta2.ResourceReference{ID: ptr.To("VPCID")}},
+			},
+		}
+		mockVPC.EXPECT().GetVPC(gomock.Any()).Return(vpcOutput, nil, nil)
+		requeue, err := clusterScope.ReconcileVPC()
+		g.Expect(err).To(BeNil())
+		g.Expect(requeue).To(BeTrue())
+	})
+
+	t.Run("When VPC ID is already set and is proper", func(t *testing.T) {
```
Comment (Karthik-K-N) [inline cloud/scope/powervs_cluster_test.go:1864]:
```suggestion
	t.Run("When VPC ID is already set and is valid", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go (hunk 5 above)
Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go]: so since you are only setting the status, You can ignore that check.

**Review by Karthik-K-N — COMMENTED** [review]
/lgtm

/assign @Amulyam24 @Prajyot-Parab

File: cloud/scope/powervs_cluster_test.go [hunk 11]
```diff
@@ -1704,3 +1704,521 @@ func TestCreateServiceInstance(t *testing.T) {
...
+			IBMVPCClient:      mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{Spec: infrav1beta2.IBMPowerVSClusterSpec{VPC: &infrav1beta2.VPCResourceReference{ID: ptr.To("VPCID")}}},
+		}
+		vpcOutput := &vpcv1.VPC{Name: ptr.To("VPCName"), ID: ptr.To("VPCID"), DefaultSecurityGroup: &vpcv1.SecurityGroupReference{ID: ptr.To("DefaultSecurityGroupID")}}
+		mockVPC.EXPECT().GetVPC(gomock.Any()).Return(vpcOutput, nil, nil)
+		vpcID, err := clusterScope.checkVPC()
+		g.Expect(err).To(BeNil())
+		g.Expect(vpcID).To(Equal(*vpcOutput.ID))
+	})
+	t.Run("When Spec VPC ID is not set and GetVPCByName returns success", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:2177]:
```suggestion
	t.Run("When spec.VPC.ID is not set and GetVPCByName returns success", func(t *testing.T) {
```
Please change wherever applicable

File: cloud/scope/powervs_cluster_test.go [hunk 12]
```diff
@@ -1704,3 +1704,521 @@ func TestCreateServiceInstance(t *testing.T) {
...
+
+	setup := func(t *testing.T) {
+		t.Helper()
+		mockCtrl = gomock.NewController(t)
+		mockVPC = mock.NewMockVpc(mockCtrl)
+	}
+	teardown := func() {
+		mockCtrl.Finish()
+	}
+	t.Run("When resourceGroupID is nil", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1896]:
```suggestion
	t.Run("When resourceGroupID is not set", func(t *testing.T) {
```

File: cloud/scope/powervs_cluster_test.go [hunk 13]
```diff
@@ -1704,3 +1704,521 @@ func TestCreateServiceInstance(t *testing.T) {
...
+		clusterScope := PowerVSClusterScope{
+			IBMVPCClient:      mockVPC,
+			IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{},
+		}
+
+		vpcID, err := clusterScope.createVPC()
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(vpcID).To(BeNil())
+	})
+	t.Run("When resourceGroupID is not nil and create VPC is successful", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1910]:
```suggestion
	t.Run("When resourceGroupID is set and create VPC is successful", func(t *testing.T) {
```
Please change wherever applicable.

**Review by Amulyam24 — COMMENTED** [review]
Minor suggestions, LGTM!

**Comment by Shilpa-Gokul** [comment]
@Prajyot-Parab @Amulyam24 @Karthik-K-N Have addressed all the comments, please take a look

**Review by Amulyam24 — APPROVED** [review]
/lgtm

/cc @Prajyot-Parab

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

**Pull request merged**